### DISCUSSION TITLE
Insights: Most viewed items and similar

### DISCUSSION THREAD

**Discussion by kropsi** [directus/directus#8479] [Feature Requests]
Introduction of "Insights" module made me wonder, how could we show most viewed or most searched items.

It makes total sense that gathering all the datapoints for every collection will be an overkill, but it might be usefull to be able to turn it on for the selected ones.

The other approach could maybe also be introduction of trigger actions like "Viewed" or "Searched", so maybe you could trigger some custom functionality to gather this information using webhooks.

Any other ideas?

**Comment by joselcvarela** [comment]
Hello @kropsi,
Maybe it will be great to do this in an easier way, but I think this could be possible today.

You can create `Searches` and `Visits` models and insert data into this models manually by registering a custom hook (not tested):
```js
// extensions/hooks/searches/index.js:

module.exports = function registerHook({ services, exceptions }) {
	const { ItemsService } = services;

	return {
		'middlewares.init.after': async function ({ router }) {
		  const searches = new ItemsService('searches')

			router.use(async (req,res,next) => {
				const {search} = req.query

				await searches.create({
					term: search
				})

				next();

			})
			
		},
	};
};

**Reply by kropsi** [reply]
Thank you @joselcvarela , I will check if it out.

**Comment by rijkvanzanten** [comment]
That's an interesting idea. The current aggregation setup doesn't work cross-collection, so it's a little tricky to "enable" within the app/api. (Also not sure how you'd do that in SQL in the first place, as you'd always use a single table as an entry point, though @Oreilles might have more thoughts not that)

**Reply by oreilles** [reply]
Directus already stores activity for `authenticate`, `create`, `update` and `delete` for every collection. It could very well store `read` query as well. Then getting most viewed items would be as simple as
```sql
SELECT collection, item, count(*)
FROM directus_activity
WHERE action = 'read'
GROUP BY collection, item
ORDER BY collection, count desc
```

**Reply by rijkvanzanten** [reply]
> Directus already stores activity for `authenticate`, `create`, `update` and `delete` for every collection. It could very well store `read` query as well. Then getting most viewed items would be as simple as
> 
> ```sql
> 
> SELECT collection, item, count(*)
> 
> FROM directus_activity
> 
> WHERE action = 'read'
> 
> GROUP BY collection, item
> 
> ORDER BY collection, count desc
> 
> ```

Yes, though.. I'm worried we'd blow up that table as the number of read operations often greatly outnumbers the writes

**Reply by alexbrowng** [reply]
If it is allowed to configure at the collection level? Extend the current functionality to mark which level of activity you want to save, create, update, delete and/or read.

**Reply by gensmann** [reply]
> If it is allowed to configure at the collection level? Extend the current functionality to mark which level of activity you want to save, create, update, delete and/or read.

and maybe make it role dependent, so it would be possible to make a role which reads or writes a lot but doesn't have to get tracked, to save load on the db. (like an external import/export api role)?

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
